### DISCUSSION TITLE
다른 유저의 팔로워 목록 조회 시 내가 팔로워 유저를 팔로잉하고 있는지 여부 확인 방법

### DISCUSSION THREAD

**Discussion by woo-chang** [Busan-Dinosaur/Foodbowl-Backend#139] [의논해볼까요?]
다른 유저의 팔로잉, 팔로워 목록 조회 시 내가 팔로잉, 팔로워 목록에 있는 유저를 팔로잉 하고 있는지 여부를 나타내야 합니다. 이때 팔로잉 하고 있는지 여부를 확인할 수 있는 방법이 2가지가 존재합니다.

## 1. 나의 모든 팔로잉 목록을 조회

나의 모든 팔로잉 목록을 조회해서 확인해야하는 유저가 나의 팔로잉 목록에 포함되어 있는지 여부를 확인하면 됩니다.

### 장점

- 모든 팔로잉 목록 조회 쿼리 1번 이후로는 불러와져있는 리스트를 통해 나머지 회원들도 확인할 수 있습니다.
- 아래의 코드와 같이 객체지향적으로 코드를 작성할 수 있습니다.

### 단점

- 추후에 인스타그램처럼 팔로잉 수가 어마어마하게 많아진다면 모든 팔로잉 데이터를 메모리에 불러와야하기 때문에 성능적인 이슈가 발생할 수 있습니다.

```java
public static OtherUserFollowerResponse of(Member follower, Member searchMember) {
  return new OtherUserFollowerResponse(
    follower.getId(),
    follower.getProfileImageUrl(),
    follower.getNickname(),
    follower.getFollowerCount(),
    searchMember.isFollowing(follower)
  );
}
```

## 2. 각 유저마다 팔로잉 여부를 조회

나의 모든 팔로잉 목록을 불러오는 것이 아닌 각각의 유저마다 쿼리를 통해 확인하는 방법입니다.

### 장점

- 팔로잉 수가 많더라도 페이징된 유저의 수만큼 데이터를 불러오면 됩니다.

### 단점

- 팔로잉 수가 적은 상황에서는 페이징 수만큼 쿼리에 의한 네트워크 비용이 큽니다.
- 도메인 모델 패턴과는 거리가 멀어질 수 있는 코드를 작성하게 됩니다.

---

위와 같은 상황에서 어떤 코드를 작성해야할지 고민하게 되었는데 같이 얘기해보면 좋을 것 같아요 :)

**Comment by Kim0914** [comment]
깊게 고민하신 부분 너무 좋습니다 :)

1번 방법의 경우에 팔로잉 수가 백 만, 천 만 이상인 경우에 성능에 영향을 미칠 수 있을 것 같네요 !

말씀해주신 `어마어마하게 많아진다면`의 기준이 어느정도의 숫자인지 궁금해요 😀

다만 SNS 특성상 팔로워 수가 팔로잉 수보다 압도적으로 많기 때문에, 팔로잉 수가 엄청나게 많은 경우는 별로 없을 것 같긴 합니다.
(보통 셀럽인 계정의 팔로워가 많은데, 팔로잉은 잘 하지 않음. 반대로 팔로잉을 많이 하는 계정은 극히 드뭄)

즉, 나의 팔로워가 다른 계정을 팔로우하는지의 여부는 표시하지 않습니다.

이런 상황에서 한번에 내 팔로잉 목록을 메모리로 불러오는 것이 부담이 되는 상황이라면, 2번 방법을 적용해볼 수 있을 것 같아요.
대신 페이징 수 만큼 네트워크를 타는 비용은 어쩔 수 없는 부분인 것 같구요. 
네트워크 비용 때문에 1번 방법처럼 한번에 벌크로 가져오는 경우가 있을 것 같네요.

2번 방법에서 궁금한 점이 있는데, 각각의 유저마다 쿼리를 통해 확인한다면 데이터베이스 내부에서는 내부 테이블을 full scan하지 않나요?!
20 명을 한 페이징해서 가져오고, 내가 팔로잉하는 유저가 300명이라면 최악의 상황을 고려했을 때 20 * 300번..? (이 부분은 그냥 추측입니다. 자세히는 모르겠어용)

그렇다면 DB 내부에서 풀 스캔하는 것 보다, 자바 단에서 한번에 불러온 후 탐색하는게 비용이 적지 않을까요??

저도 조금 헷갈려서 ㅎㅎ 의견 남겨주시면 감사하겠습니다 !

**Reply by woo-chang** [reply]
`JVM`은 실행되는 환경에 따라 힙 메모리의 최대 사이즈를 정한다고 해요. 힙 메모리는 설정하지 않으면 해당 환경 메모리의 `1/4`로 지정하게 되는데, 현재 개발 서버의 메모리는 `16GB`이기때문에 `4GB`정도 할당될 것으로 생각합니다.

GPT 선생님에 따르면 `Member` 객체 하나당 대략 500바이트의 메모리를 사용한다고 합니다. 만명이라면 500B x 10000 = 5000KB = 5MB정도로 괜찮을 것 같지만, 100만 팔로워가 생긴다면 조금 고려해봐야할 문제가 될 것 같아요! 500MB = 0.5GB이기 때문에 8명이 동시에 조회를 해버리는 순간 `Out Of Memory` 문제가 발생할 가능성이 생기기 때문입니다.

현재는 시스템 초기라 괜찮을 것 같지만 사용자가 늘어남에 따라 계속해서 고민이 필요한 부분이라고 생각해요 :)

2번 방법을 사용하면 `Follow` 테이블에서 `following_id`, `follower_id`로 조회를 하게되는데, 두 컬럼은 유니크로 묶여 있기에 인덱스가 걸려 있어요. 따라서 DB 풀 스캔이 아닌 인덱스로 찾게될 것 같아서 생각하시는만큼 비용이 들지 않을 것 같습니다!

현재는 어떤 방법을 사용하든 괜찮아보이고, 선택하면 될 것 같은데 어떤 것이 현재 상황에서 최선의 선택일까요? 🤔

**Reply by Kim0914** [reply]
흠... 저는 선택하기가 참 어렵네요 😂
객체지향적인 코드를 가져가고 싶은 욕심이 있어서 1번 방법이 마음에 드는데, 성능과 추후 비용을 생각하면 2번 방법이 더 적절해보이네요.

사실 프로젝트 초기이기 때문에 1번으로 구현해도 크게 문제는 없지만, 추후 서버를 늘리는 상황에서 비용적으로 부담이 될 수 있으므로 2번으로 진행하는게 좋아보여요 ㅎㅎ
